shared/dataflow/codeql/dataflow/DataFlow.qll

```diff
@@ -851,4 +851,231 @@ module DataFlowMake<LocationSig Location, InputSig<Location> Lang> {
+      InputPathNode arg, InputPathNode param, InputPathNode ret, InputPathNode out
+    );
+
+    /**
+     * Performs a forward or backward pass computing which `(node, toString)` pairs can subsume their corresponding
+     * path nodes.
+     *
+     * This is similar to automaton minimization, but for an NFA. Since minimizing an NFA is NP-hard (and does not have
+     * a unique minimal NFA), we operate with the simpler model: for a given `(node, toString)` pair, either all
+     * corresponding path nodes are merged, or none are merged.
+     *
+     * Comments are written as if this checks for outgoing edges and propagates backward, though the module is also
+     * used to perform the opposite direction.
+     */
+    private module MakeDiscriminatorPass<
+      collapseCandidateSig/2 collapseCandidate, stepSig/4 step, subpathStepSig/4 subpathStep>
+    {
+      /**
+       * Gets the number of `(key, val, node, toString)` tuples reachable in one step from `pathNode`.
+       *
+       * That is, two edges are counted as one if their target nodes are the same after projection, and the edges have the
+       * same `(key, val)`.
+       */
+      private int getOutDegreeFromPathNode(InputPathNode pathNode) {
+        result =
+          count(Node node, string toString, string key, string val |
+            step(pathNode, getAPathNode(node, toString), key, val)
+          )
+      }
+
+      /**
+       * Gets the number of `(key, val, node2, toString2)` pairs reachable in one step from path nodes corresponding to `(node, toString)`.
+       */
+      private int getOutDegreeFromNode(Node node, string toString) {
+        result =
+          strictcount(Node node2, string toString2, string key, string val |
+            step(getAPathNode(node, toString), getAPathNode(node2, toString2), key, val)
+          )
+      }
+
+      /**
+       * Like `getOutDegreeFromPathNode` except counts `subpath` tuples.
+       */
+      private int getSubpathOutDegreeFromPathNode(InputPathNode pathNode) {
+        result =
+          count(Node n1, string s1, Node n2, string s2, Node n3, string s3 |
+            subpathStep(pathNode, getAPathNode(n1, s1), getAPathNode(n2, s2), getAPathNode(n3, s3))
+          )
+      }
+
+      /**
+       * Like `getOutDegreeFromNode` except counts `subpath` tuples.
+       */
+      private int getSubpathOutDegreeFromNode(Node node, string toString) {
+        result =
+          strictcount(Node n1, string s1, Node n2, string s2, Node n3, string s3 |
+            subpathStep(getAPathNode(node, toString), getAPathNode(n1, s1), getAPathNode(n2, s2),
+              getAPathNode(n3, s3))
+          )
+      }
+
+      /** Gets a successor of `node` including subpath flow-through. */
+      InputPathNode stepEx(InputPathNode node) {
+        step(node, result, _, _)
+        or
+        subpathStep(node, _, _, result) // assuming the input is pruned properly, all subpaths have flow-through
```

Comment:
This should already be present in `step`, so it should be superfluous here, but it cannot hurt, so I don't mind keeping it. But maybe add a comment about this.